pm/sendermonitor.go

```diff
@@ -272,17 +263,103 @@ func (sm *senderMonitor) startCleanupLoop() {
 
 // cleanup removes tracked remote senders that have exceeded
 // their ttl
-func (sm *senderMonitor) cleanup() {
+func (sm *LocalSenderMonitor) cleanup() {
 	sm.mu.Lock()
 	defer sm.mu.Unlock()
 
 	for k, v := range sm.senders {
 		if unixNow()-v.lastAccess > int64(sm.ttl) {
 			// Signal the ticket queue consumer to exit gracefully
 			v.done <- struct{}{}
-
+			v.subScope.Close() // close the maxfloat subscriptions
 			delete(sm.senders, k)
 			sm.smgr.Clear(k)
 		}
 	}
 }
+
+func (sm *LocalSenderMonitor) redeemWinningTicket(ticket *SignedTicket) error {
+	maxFloat, err := sm.MaxFloat(ticket.Ticket.Sender)
+	if err != nil {
+		return err
+	}
+
+	// if max float is zero, there is no claimable reserve left or reserve is 0
+	if maxFloat.Cmp(big.NewInt(0)) <= 0 {
+		if err := sm.QueueTicket(ticket); err != nil {
+			return err
+		}
+		return errors.New("max float is 0")
+	}
+
+	// If max float is insufficient to cover the ticket face value, queue
+	// the ticket to be retried later
+	if maxFloat.Cmp(ticket.Ticket.FaceValue) < 0 {
+		if err := sm.QueueTicket(ticket); err != nil {
+			return err
+		}
+		return fmt.Errorf("insufficient max float sender=%v faceValue=%v maxFloat=%v", ticket.Ticket.Sender.Hex(), ticket.Ticket.FaceValue, maxFloat)
+	}
+
+	// Subtract the ticket face value from the sender's current max float
+	// This amount will be considered pending until the ticket redemption
+	// transaction confirms on-chain
+	sm.subFloat(ticket.Ticket.Sender, ticket.Ticket.FaceValue)
+
+	defer func() {
+		// Add the ticket face value back to the sender's current max float
+		// This amount is no longer considered pending since the ticket
+		// redemption transaction either confirmed on-chain or was not
+		// submitted at all
+		//
+		// TODO(yondonfu): Should ultimately add back only the amount that
+		// was actually successfully redeemed in order to take into account
+		// the case where the ticket was not redeemd for its full face value
+		// because the reserve was insufficient
+		if err := sm.addFloat(ticket.Ticket.Sender, ticket.Ticket.FaceValue); err != nil {
+			glog.Error(errors.Wrap(err, "error adding to max float"))
```

Comment:
Right now we're logging an error in `addFloat()` and here as well.

We should either:
- Log the error in `addFloat()` and not handle an error here

or

- Return an error in `addFloat()` and log the error here

Reply:
Oh yes, now returning an error again from addFloat and logging it in the caller dc0e6020